linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala

```diff
@@ -97,28 +107,123 @@ private object MaxModuleAnalyzer {
   }
 
+  /** Tagger performs the actual grouping of classes into modules.
+   *
+   *  To group classes into modules appropriately, we want to know for
+   *  each class, "how" it can be reached. In practice, this means we
+   *  record the path from the original public module and every
+   *  dynamic import hop we made.
+   *
+   *  Of all these paths, we only care about the "simplest" ones. Or
+   *  more formally, the minimum prefixes of all paths. For example,
+   *  if a class is reachable by the following paths:
+   *
+   *  - a -> b
+   *  - a -> b -> c
+   *  - d -> c
+   *  - d
+   *
+   *  We really only care about:
+   *
+   *  - a -> b
+   *  - d
+   *
+   *  Because if we reach the class through a path that goes through
+   *  `c`, it is necessarily already loaded.
+   *
+   *  Once we have obtained this minimal set of paths, we use the last
+   *  elements of the paths to determine the final module
+   *  grouping. This is because these have an actual static dependency
+   *  on the node in question.
+   *
+   *  Merging these tags into a single `ModuleID` is delegated to the
+   *  caller.
+   */
   private final class Tagger(infos: ModuleAnalyzer.DependencyInfo) {
-    private[this] val allTags =
-      mutable.Map.empty[ClassName, mutable.Set[ModuleID]]
+    private[this] val allPaths =
+      mutable.Map.empty[ClassName, Paths[ModuleID, ClassName]]
 
-    def tagAll(): scala.collection.Map[ClassName, scala.collection.Set[ModuleID]] = {
+    def tagAll(): scala.collection.Map[ClassName, (Set[ModuleID], Set[ClassName])] = {
       tagEntryPoints()
-      allTags
+      allPaths.map { case (className, paths) => className -> paths.ends }
     }
 
-    private def tag(className: ClassName, moduleID: ModuleID): Unit = {
-      val perClassTags = allTags.getOrElseUpdate(className, mutable.Set.empty)
-      if (perClassTags.add(moduleID))
-        infos.classDependencies(className).foreach(tag(_, moduleID))
+    private def tag(className: ClassName, pathRoot: ModuleID, pathSteps: List[ClassName]): Unit = {
+      val updated = allPaths
+        .getOrElseUpdate(className, Paths.empty)
+        .put(pathRoot, pathSteps)
+
+      if (updated) {
+        val classInfo = infos.classDependencies(className)
+        classInfo
+          .staticDependencies
+          .foreach(tag(_, pathRoot, pathSteps))
+
+        classInfo
+          .dynamicDependencies
+          .foreach(c => tag(c, pathRoot, pathSteps :+ c))
+      }
     }
 
     private def tagEntryPoints(): Unit = {
       for {
         (moduleID, deps) <- infos.publicModuleDependencies
         className <- deps
       } {
-        tag(className, moduleID)
+        tag(className, moduleID, Nil)
+      }
+    }
+  }
+
+  /** Set of shortest, mutually prefix-free paths.
+   *
+   *  Would like to make this AnyVal, but that throws Erasure on 2.11 in an
+   *  infinite loop.
+   */
+  private final class Paths[H, T] private (
+      private val content: mutable.Map[H, Paths[T, T]]) {
```

Comment:
While this recursive `Paths[H, T]` that contains `Paths[T, T]` is quite clever, I am not sure it holds its weight against two separate classes `HeadPaths` and `TailPaths` with hard-coded `ModuleID` and `ClassName`.
* The `ends` method could be implemented in a more efficient way by not creating a builder at each recursion step (the `TailPaths` class' equivalent of `ends` could take the `Builder[ClassName]` as parameter rather than creating a new one each time)
* `isEmpty` is only necessary for `TailPaths`

`put` would have to be duplicated, though.

I am not sure. Just putting this out there for consideration.

Reply:
I have tried this. I couldn't find a way of doing so without duplicating `put` and the non-trivial parts of `ends`. Basically, the little logic that distinguishes empty entries and missing entries feels non trivial with a typical interface to a `Map`. If you feel nevertheless this should be simplified, I can try harder.